ProteoformSuiteInternal/Proteoform.cs

```diff
@@ -52,62 +49,139 @@ public List<ExperimentalProteoform> identify_connected_experimentals()
                 {
-                    if (deltaM >= m.monoisotopicMass - mass_tolerance && deltaM <= m.monoisotopicMass + mass_tolerance 
-                        && (best_addition == null || Math.Abs(deltaM - m.monoisotopicMass) < Math.Abs(deltaM - best_addition.monoisotopicMass))
-                        || best_addition != null && Math.Abs(deltaM - m.monoisotopicMass) == Math.Abs(deltaM - best_addition.monoisotopicMass) && m.modificationType == "Unlocalized")
-                    {
-                        best_addition = m;
-                    }
-
-                    if (this.ptm_set.ptm_combination.Select(ptm => ptm.modification).Contains(m) 
-                        && deltaM >= -m.monoisotopicMass - mass_tolerance && deltaM <= -m.monoisotopicMass + mass_tolerance
-                        && (best_loss == null || Math.Abs(deltaM - (-m.monoisotopicMass)) < Math.Abs(deltaM - (-best_loss.monoisotopicMass))))
+                    bool within_loss_tolerance = deltaM >= -m.monoisotopicMass - mass_tolerance && deltaM <= -m.monoisotopicMass + mass_tolerance;
+                    bool can_be_removed = this.ptm_set.ptm_combination.Select(ptm => ptm.modification).Contains(m);
+                    bool better_than_current_best_loss = best_loss == null || Math.Abs(deltaM - (-m.monoisotopicMass)) < Math.Abs(deltaM - (-best_loss.monoisotopicMass));
+                    if (can_be_removed && within_loss_tolerance && better_than_current_best_loss)
                     {
                         best_loss = m;
                     }
                 }
 
                 // If they're the same and someone hasn't labeled 0 difference with a "ModificationWithMass", then label it null
-                if (best_addition == null && best_loss == null && Math.Abs(r.peak_center_deltaM) <= mass_tolerance)
+                if (best_addition == null && best_loss == null && Math.Abs(r.peak.peak_deltaM_average) <= mass_tolerance)
                 {
-                    lock (r) lock (e) assign_pf_identity(e, this, ptm_set, r, null);
+                    lock (r) lock (e) assign_pf_identity(e, this, ptm_set, r, sign, null);
                     identified.Add(e);
                 }
 
                 if (best_addition == null && best_loss == null)
                     continue;
 
                 PtmSet with_mod_change = best_loss != null ?
-                    new PtmSet(new List<Ptm>(this.ptm_set.ptm_combination.Except(this.ptm_set.ptm_combination.Where(ptm => ptm.modification.Equals(best_loss))))) :
-                    new PtmSet(new List<Ptm>(this.ptm_set.ptm_combination).Concat(new Ptm[] { new Ptm(-1, best_addition) }).ToList());
-                lock (r) lock (e) assign_pf_identity(e, this, with_mod_change, r, best_loss != null ? best_loss : best_addition);
+                    new PtmSet(new List<Ptm>(this.ptm_set.ptm_combination.Where(ptm => !ptm.modification.Equals(best_loss)))) :
```

Comment:
These represent PTM losses. It favors removing a PTM from the ptm set over adding one. The amino acid degradation would be included in the "added" PTM set as a missing AA.